### ISSUE TITLE
Ugly tables in markdown pages

### ISSUE THREAD

**Issue by mbraakhekke** [Fortran-FOSS-Programmers/ford#373]
Tables in in the markdown pages look different from those on the overview pages of modules, procedures, etc., and rather ugly, in my view. E.g., the following table:
```
| Header1        | Header2     |
| -------------- | ----------- |
| some text      | some text   |
| some text      | some text   |
```

gets rendered as:

![image](https://user-images.githubusercontent.com/10126994/147147086-691bab07-5ed7-439a-8709-7f178a9c8343.png)

The problem is that python-markdown [does not allow](https://github.com/Python-Markdown/markdown/issues/312) for adding classes to tables. I propose post-processing the HTML code of markdown tables (and possibly other elements) in FORD to assure a consistent look. This can probably done more easily after #357 gets implemented.

**Comment by ZedThree** [comment]
Possibly a very naive question, but is this "just" a matter of CSS? That is, do we need to change the CSS for tables on the markdown pages?
Or perhaps put another way, do we know why the tables look like that in the first place?

**Comment by urbanjost** [comment]
This really impacts the usefulness of markdown and is because the CSS styles completely collapse all the table elements.
Is there some reason ford does this?

**Comment by ZedThree** [comment]
@urbanjost @mbraakhekke Could you try out #384 and let me know if that's suitable?

**Comment by urbanjost** [comment]
cloned and went to branch pr384 and ran install.py and see no difference in the output. Perhaps not trying it out correctly?

**Comment by ZedThree** [comment]
Please try the following steps:

```
git clone git@github.com:Fortran-FOSS-Programmers/ford.git
cd ford
git switch fix-table-css
python3 -m venv venv
source venv/bin/activate
pip install .
```

Then `ford --version` should report as `ford version 6.1.9.dev1+g66c3d33`

**Comment by urbanjost** [comment]
better in horizontal spacing, but tried several reformats and
could not get something where multi-line descriptions worked as expected.
Every line is treated as a row, which is not what is intended. So using this
how would you create something like the following table? It still has to look
reasonable as comments in the Fortran source, so do not want to put an 
entire description on one line, as in value "4" in the following table. 


```text

 value  | descriptions
------- | -----------------------------------------------------------
      1 | means normal computation of output values of y(t) at
        | t = TOUT (by overshooting and interpolating).
      2 | means take one step only and return.
      3 | means stop at the first internal mesh point at or
        | beyond t = TOUT and return.
      4 | means normal computation of output values of y(t) at
        | t = TOUT but without overshooting t = TCRIT.
        | TCRIT must be input as RWORK(1).  TCRIT may be equal to
        | or beyond TOUT, but not behind it in the direction of
        | integration.  This option is useful if the problem
        | has a singularity at or beyond t = TCRIT.
      5 | means take one step, without passing TCRIT, and return.
        | TCRIT must be input as RWORK(1).
```

**Comment by ZedThree** [comment]
I'm afraid that's a limitation of the table syntax. See this SO answer by the maintainer of Python-markdown (the implementation that we use in ford): https://stackoverflow.com/a/57166923
I've had a quick look and can't see any extensions that offer a different table syntax.

For that particular example, you may be better off just formatting it as a list.

**Comment by urbanjost** [comment]
Yes, but that "worked" in the previous version and in other markdown versions because everything was packed with no additional padding or borders vertically and not highlighted.  SO if you try that table in the current release it looks reasonable vertically, but is packed or smashed together horizontally. My only major issue with the previous version was the horizontal layout; and in CSS they are controllable individually.

**Comment by ZedThree** [comment]
Ok, you probably want to play about with `local.css`, for example adding:

```css
td, th {
    padding-right: 10px;
}
```

**Comment by urbanjost** [comment]
That looks like exactly what I need; and perhaps just changing the defaults there is all I need. I edited the local.css file and was able to get it to look like I wanted with you example and a few other changes. But I am having trouble finding the documentation for local.css. It gets rebuilt each time I run ford(1).  I now see the -css switch on ford(1); and am assuming if I place either a copy of local.css modified or just my changes they get added and can play with that assumption but so far I have not found where this is documented. Any hints on how to use this in the documentation that I am missing? This indeed looks like with a little bit of work with CSS this will fix my issues; but I do think the defaults should be changed, perhaps just be adding what you provided to the defaults.  Just adding that made the default behavior much better.  Thanks!

**Comment by mbraakhekke** [comment]
Sorry for the late reply from my side. I did a quick test and it looks good. The example table I posted before now looks like this:
![image](https://user-images.githubusercontent.com/10126994/153281444-adf4c69b-f688-4c7e-a020-7fa384db7cd3.png)
I'll do some more testing and report back.

**Comment by mbraakhekke** [comment]
After some further test I think that the new tables look great now. This issue can be closed for me.

**Comment by ZedThree** [comment]
@mbraakhekke Do you mean with #384, or just the CSS in local.css as in my comment above?

If you meant #384, would you be ok if there was an option to enable this? Maybe something like `zebra_tables`

**Comment by mbraakhekke** [comment]
Yeah I meant #384.

With regard to adding an option: it's not really my call but I can't imagine why anyone would not want to use it given how the markdown tables look without it. So if you want to add an option I would make the default `true`/`yes`.
Also, what exactly would this option mean? That markdown tables are formatted in the same manner as the auto-generated tables? If so, I wonder if  `zebra_tables` would be the right name, given that it pertains specifically to markdown tables. Moreover, I could imagine in the future functionality would added to allow different formatting of tables, in which case `zebra` may not be appropriate any more.

By the way: thanks for fixing this! :clap:

**Comment by ZedThree** [comment]
Ok, #384 now has the minimal fix, and an optional extension for the striped tables. I forgot that we already have the functionality to enable markdown extensions, so you can use that.

Another option is to use the [md_in_html](https://python-markdown.github.io/extensions/md_in_html/) extension to wrap your table in a `<div>` tag and then set some CSS for that:

in `docs.md`:

```markdown
md_extensions: markdown.extensions.md_in_html
css: user.css
---

<div class="my-table">
| Header1 | Header2 |
| some text | other text |
</div>
```

and in `user.css`:

```css
.my-table > table > tbody > tr:nth-child(odd) {
    background-color: #f9f9f9;
    /* other CSS styling */
}
```

Lastly, for even more control over tables without resorting to html, is [markdown-grid-tables](https://pypi.org/project/markdown-grid-tables/) which enables reStructuredText-style tables. You might still want to use the above `md_in_html` trick to apply your own CSS to the tables.
